### PR TITLE
ReaderBookmark: save bookmark page string

### PR DESCRIPTION
**Pull request by hius07**
Bookmark page numbers for reflowable documents are not actual page numbers but some strings useless for users.
Exporter plugin and external programs that deal with our bookmarks (like KOHighlights) cannot get correct bookmark page number if a document layout has been changed.

This PR allows **manual** updating the page numbers and saving them to the book sdr file in bookmark property `page_string`.
Exporter plugin takes this string if it exists and put it as "page" in export results.
External programs can add support of it if they will.

Please note that we do not maintain this property **automatically**. Newly created bookmarks do not have it.

Pinging developers and heavy highlighters @noembryo, @isosphere, @mergen3107, @offset-torque, @ichnilatis-gr.
Is this feature useful?

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11464)

### CODE REVIEW

**Comment by noembryo** [comment]
This looks like a really neat feature.

What I would like to clarify though, is this.
From what I can see from my code, the page numbers are the keys of the `"highlight"` object of the `sdr` metadata.
Does this new feature renumbers these keys or it adds a new value with the new string?
If its the former, no change has to be made to my code, so, yoo-hoo! Free lunch.
If it's the latter, it might make things more complicated, because then we will have a mismatch..

Also, why is this not made automatically when the layout has changed.
Of course it must be delayed, so not many changes will take place while someone is trying to find the optimal layout.
Perhaps, not so many users will use it otherwise..

**Comment by ichnilatis-gr** [comment]
Thank you for this feature. I also had thought of this problem. I think is useful. Please explain about the formats (linear, non-linear etc.).

**Comment by hius07** [comment]
These new page numbers (actually, strings) are the members of the `bookmarks` object.
They do not replace old page numbers.
If new `page_string` exists in sdr, we can assume that a user has generated it recently, it is correct and it should override page numbers from other sources (like `highlight` object).

"linear" - usual page numbers ("100")
"non-linear" - for documents with non-linear fragments ("[1]100")
"label" - for documents with reference pages
"none" - remove this property from bookmarks

**Comment by noembryo** [comment]
>"non-linear" - for documents with non-linear fragments ("[1]100")
"label" - for documents with reference pages

These two might create a problem, since in my code the sorting is happening after a conversion to `int`.
Depending on the format it might need something more than an alphanumeric sorting.
Is there a standard for it?
>"none" - remove this property from bookmarks

Perhaps naming this "clear" is clearer? 😉

**Comment by noembryo** [comment]
Also, why not just renumbering the keys?

**Comment by hius07** [comment]
> Is there a standard for it?

non-linear
https://github.com/koreader/koreader/blob/93d1cfe41d809bebaa6637ba48d3ded5f1cab80a/frontend/apps/reader/modules/readerbookmark.lua#L1525
labels
https://github.com/koreader/koreader/blob/8010808a1fa0460c9e2c01fc2c4c6f23ba11063a/frontend/apps/reader/modules/readerpagemap.lua#L118
I'm afraid, cannot be alpha-sorted.
Please note that the `bookmarks` object is perfectly sorted not only by the page number, but also by the position in the page.

**Comment by offset-torque** [comment]
So this is what I understood:
1. We don't (can't) have page numbers in reflowable documents because there is no ground truth, rendering changes page numbers (which makes bookmark page numbers useless)
2. This proposed feature (manually) updates "bookmark page numbers" after a layout change
3. After this renumbering, exported page numbers **will match** the current KOReader page numbers
4. If the user changes the layout after invoking this feature, KOReader<->exported page numbers **will get desynchronized** so he has to re-update page bookmark strings by invoking this again.

Did I misunderstand anything ?

**Comment by poire-z** [comment]
> In our Bookmarks window we always have correct page numbers, so this new feature can be useful for export or external programs only

Why not just auto-update the new property/string when that bookmark window is launched (as you have to get the updated value to display it in that window) ?
And why not auto-choose the format the user has decided to use/show (in the footer, in that bookmark list or ToC) ?
Or put all available variations with each a different key ? `reference_page`, `screen_page`, `flow_page` ?

> Depending on the format it might need something more than an alphanumeric sorting.
Is there a standard for it?

@noembryo : I think you should consider them just as additional tags, and keep using whatever you use for associations/finding/sorting.
The most obvious case is "Reference page numbers" / "labels": these are the page numbers of the hardcopy paper books - the user may prefer to have them in his exports as he can then rematch them to the paper version of the book.
These are not always numbers, ie. `Page XVI` in some introduction in a book. So,they should just be considered as random strings attached to a bookmark, for display only, not for any technical use.

File: frontend/apps/reader/modules/readerbookmark.lua [hunk 1]
```diff
@@ -1470,4 +1480,63 @@ function ReaderBookmark:getHighlightDrawer(datetime)
...
+        {_("non-linear"), "nonlinear"},
+        {_("label"), "label"},
+        {_("none"), "none"},
+    }
+    local radio_buttons = {}
+    for _, v in ipairs(types) do
+        local text, type = unpack(v)
+        local enabled = true
+        if type == "nonlinear" then
+            enabled = (self.ui.rolling and self.ui.document:hasHiddenFlows()) and true or false
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerbookmark.lua:1495]: Not with the recent "custom hidden flows" also available with PDFs ?

**Comment by isosphere** [comment]
I believe this is a useful feature, but I feel that the user experience workflow will make it of limited practical use. If a user must manually open the bookmarks menu, open the filter menu, and click this update button, and do this per-book of a particular format, it will see very limited usage. That's a lot of clicking to update a library. 

If it could be done across the entire library (maybe that's what's implemented here and I'm mistaken), that would be low friction and of great utility. Or, if they could specify that they want them automatically updated on a layout change or if the field is null at the cost of a bit of lag, that's something I'd personally select - but I'm using an Android phone which has a lot more power than a Kobo.

**Comment by offset-torque** [comment]
I agree with **isosphere**'s and **poire-z**'s suggestions. UX need to carefully decided in this case. 

The essence of this feature is to provide a reference to the exported bookmarks. So users can see the highlight in the exported medium and open that page in KOReader. This means book will stay in the device as the reference in the last exported form (in a kind of frozen state).

Problem is, if user changes any setting, exported page numbers will be practically useless. So we have to prevent the user shooting themselves in the foot. He changes the margins, font weight, line spacing...page numbers gone. 

At least there should be a warning for any option that might shift these page numbers: **"If you change this setting your exported page numbers will not be valid anymore and you will have to update and export again"** (This sounds like so much effort for all edge cases etc.)

Otherwise users has to keep mental note of which documents they exported so they can avoid changing anything in them. In a worst case scenario I can think, a user who is trying different fonts, exported some of the books with the Literata font for example. Then he deletes the font from their device and all that documents fallback to default font and their exported page numbers become invalid.

In the end this proposal looks like a useful feature. Currently I export with **KoHighlights** (thanks @noembryo) and if I need to find a highlight in the book, I search part of it. Because it is very annoying to do this on an e-ink device, I need to do it on my computer. This feature will allow me to directly go to that page which is cool. But as I said, I and other users will need at least a warning as described above.

As a workflow improvement, there can be an option like **"Update bookmark page numbers when book is marked as finished"** or a menu item at the end of the book. And there can be a gesture action for this. So when KOReader notices that an already exported book has been changed and show a dialog like **"Book is re-rendered and bookmark page numbers need to be recalculated"** maybe with a button says re-calculate.

In summary this needs more effort to be effectively usable. I am personally ok with my current solution (searching the string on a computer) because I knew other solutions like this would require much effort so I never even opened an issue. The requirements I wrote above was for other users, because once they start using this feature, it becomes our responsibility to prevent them from losing data (page numbers in this case) with a simple thing like changing margins.

**Comment by hius07** [comment]
Thank you all for the discussion. I see the proposed solution is not user-friendly and must be improved. I'll think it over.
